rust/bridge/node/futures/src/future.rs

```diff
@@ -0,0 +1,205 @@
+/// This is the result of `await`ing the future.
+///
+/// Panics in the transformation function will be propagated to the `await`ing context.
+pub struct JsFuture<T: 'static + Send> {
+    // In practice there will only be one strong reference to one of these from Rust,
+    // and two weak ones from JavaScript (for `resolve` and `reject`).
+    shared: Arc<Mutex<JsFutureState<T>>>,
+}
+
+impl<T: 'static + Send> Future for JsFuture<T> {
+    type Output = T;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context) -> Poll<Self::Output> {
+        let mut state_guard = self.shared.lock().unwrap();
+        let state = mem::replace(&mut *state_guard, JsFutureState::Consumed);
+        match state {
+            JsFutureState::Complete(Ok(result)) => return Poll::Ready(result),
+            JsFutureState::Complete(Err(panic)) => resume_unwind(panic),
+            JsFutureState::Consumed => panic!("already consumed"),
+            JsFutureState::Waiting { .. } => {}
+        }
+        *state_guard = state.waiting_on(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+struct WeakFutureToken<T: Send + 'static>(Weak<Mutex<JsFutureState<T>>>);
+
+impl<T: Send + 'static> WeakFutureToken<T> {
+    /// Creates a token referencing `future`.
+    fn new(future: &JsFuture<T>) -> Self {
+        Self(Arc::downgrade(&future.shared))
+    }
+
+    /// Produces a JavaScript function value representing [fulfill_promise]
+    /// with the first argument bound to this token, boxed.
+    fn bind_fulfill_promise<'a, C: Context<'a>, R: JsPromiseResultConstructor>(
+        &self,
+        cx: &mut C,
+    ) -> JsResult<'a, JsValue> {
+        let fulfill = JsFunction::new(cx, fulfill_promise::<T, R>)?;
+        let bind = fulfill
+            .get(cx, "bind")?
+            .downcast_or_throw::<JsFunction, _>(cx)?;
+        let bind_args = vec![
+            cx.undefined().upcast::<JsValue>(),
+            cx.boxed(self.clone()).upcast(),
+        ];
+        bind.call(cx, fulfill, bind_args)
+    }
+}
+
+impl<T: Send> Finalize for WeakFutureToken<T> {}
+
+impl<T: Send> Clone for WeakFutureToken<T> {
+    fn clone(&self) -> Self {
+        Self(Weak::clone(&self.0))
+    }
+}
+
+/// Registered as the callback for the `resolve` and `reject` parameters of [`Promise.then`][then].
+///
+/// This callback assumes its first (bound) argument represents a boxed [WeakFutureToken].
+/// If the referenced [JsFutureState] is still alive, it is fulfilled and the future awoken.
+///
+/// [then]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then
+fn fulfill_promise<T: Send + 'static, R: JsPromiseResultConstructor>(
+    mut cx: FunctionContext,
+) -> JsResult<JsUndefined> {
+    let js_result = cx.argument(1)?;
+    let future = &cx.argument::<JsBox<WeakFutureToken<T>>>(0)?.0;
+
+    if let Some(future) = future.upgrade() {
```

Comment:
I considered that, but putting it here makes it very clear that the JsFuture can't be fulfilled in any *other* way. Maybe that's fine, though—the `resolve` method will still take a Context, and that means you're in a synchronous JavaScript call.

Reply:
\>< Now I remember why I went through `bound_reject` in the builder: the `transform` takes a FunctionContext to operate in, and the builder only has a TaskContext. I can't make the transform take a `dyn Context` instead because Context isn't Sized. Even then it didn't come out as nicely as I liked because it either had to return the waker or take the MutexGuard (or Mutex, or Arc) in order to unlock before waking the Future.